machine-tests/src/test/java/com/dylibso/chicory/testing/MachinesTest.java

```diff
@@ -270,13 +275,96 @@ public void shouldCallIndirectAotToInterpreter() {
+            var moduleName = instance.module().moduleName().orElse(fallbackModuleName);
+            var funcName =
+                    instance.module()
+                            .functionName(funcIdx)
+                            .map(str -> Demangler.demangle(str))
+                            .orElse("function[" + funcIdx + "]");
+            return List.of(new StackTraceElement(moduleName, funcName, null, -1));
+        };
+    }
+
+    private static BiFunction<Instance, Integer, List<StackTraceElement>>
+            demanglerExceptionConverter(WasmModule debugModule) {
+        return (instance, funcIdx) -> {
+            var moduleName = debugModule.moduleName().orElse("wasm-module");
+            var funcName =
+                    debugModule
+                            .functionName(funcIdx)
+                            .map(str -> Demangler.demangle(str))
+                            .orElse("function[" + funcIdx + "]");
+            return List.of(new StackTraceElement(moduleName, funcName, null, -1));
+        };
+    }
+
+    @Test
+    public void shouldEmitUnderstandableStackTracesWithInterpreter() throws Exception {
+        var instance =
+                Instance.builder(loadModule("compiled/count_vowels.rs.wasm"))
+                        .withExceptionConverter(MachinesTest::demanglerExceptionConverter)
+                        .build();
+        var countVowels = instance.export("count_vowels");
+        var exception = assertThrows(TrapException.class, () -> countVowels.apply(0, -1));
+        var exceptionTxt = readStackTrace(exception);
+
+        assertTrue(exceptionTxt.contains("count_vowels.wasm.count_vowels"));
+        assertTrue(exceptionTxt.contains("count_vowels.wasm.std::panicking::rust_panic_with_hook"));
+    }
+
+    @Test
+    public void shouldEmitUnderstandableStackTracesRuntimeCompiled() throws Exception {
+        var instance =
+                Instance.builder(loadModule("compiled/count_vowels.rs.wasm"))
+                        .withMachineFactory(MachineFactoryCompiler::compile)
+                        .withExceptionConverter(MachinesTest::demanglerExceptionConverter)
+                        .build();
+        var countVowels = instance.export("count_vowels");
+        var exception = assertThrows(TrapException.class, () -> countVowels.apply(0, -1));
+        var exceptionTxt = readStackTrace(exception);
+
+        assertTrue(exceptionTxt.contains("count_vowels.wasm.count_vowels"));
+        assertTrue(exceptionTxt.contains("count_vowels.wasm.std::panicking::rust_panic_with_hook"));
+    }
+
+    @Test
+    public void shouldEmitUnderstandableStackTracesBuildTimeCompiled() throws Exception {
+        var debugModule = loadModule("compiled/count_vowels.rs.wasm");
+        var instance =
+                Instance.builder(CountVowels.load())
+                        .withMachineFactory(CountVowels::create)
+                        .withExceptionConverter(demanglerExceptionConverter(debugModule))
```

Comment:
This is where the new design shines, we can inject the debugging logic using a different `WasmModule`.